server/map_rpc_server.go

```diff
@@ -263,6 +267,33 @@ func (t *TrillianMapServer) SetLeaves(ctx context.Context, req *trillian.SetMapL
 				HashedValue: l.LeafHash,
 			})
 		}
+
+		// Work around a performance issue when using the map in
+		// single-transaction mode by preloading all the nodes we know the
+		// sparse merkle writer is going to need.
+		if t.opts.UseSingleTransaction && t.opts.UseLargePreload {
+			readRev, err := tx.ReadRevision(ctx)
+			if err != nil {
+				return err
+			}
+			nidSet := make(map[string]bool)
+			nids := make([]storage.NodeID, 0, len(hkv)*256)
+			for _, i := range hkv {
+				nid := storage.NewNodeIDFromHash(i.HashedKey)
+				sibs := (&nid).Siblings()
```

Comment:
Not in scope for this PR but we could look at using *NodeIDs more widely. I suspect we could save on some slice copying.